File: netlify/functions/validate.js
const crypto = require('crypto');

const EXPIRATION_DATE = new Date(Date.UTC(2026, 0, 1)); // 2026-01-01

exports.handler = async () => {
  try {
    let privateKey = process.env.VLTR_BETA_PRIVATE_KEY;
    if (!privateKey) {
      return {
        statusCode: 500,
        body: 'Server configuration error: Private key not found.',
      };
    }

    // Handle keys stored with escaped newlines ("\n")
    if (privateKey.includes('\\n')) {
      privateKey = privateKey.replace(/\\n/g, '\n');
    }

    // Normalize line endings and trim extra whitespace
    privateKey = privateKey.replace(/\r\n?/g, '\n').trim();

    if (!privateKey.includes('BEGIN') || !privateKey.includes('END')) {
      return {
        statusCode: 500,
        body: 'Server configuration error: Invalid private key format.',
      };
    }

    // Ensure PEM has trailing newline the OpenSSL decoder expects
    if (!privateKey.endsWith('\n')) {
      privateKey += '\n';
    }

    const payload = {
      issue_date: new Date().toISOString(),
      expiration_date: EXPIRATION_DATE.toISOString(),
      issuer: 'VLTR Beta Program',
    };

    const payloadJson = JSON.stringify(payload);
    const signer = crypto.createSign('sha256');
    signer.update(payloadJson);
    signer.end();

    const keyObject = crypto.createPrivateKey({ key: privateKey, format: 'pem', type: 'pkcs8' });
    const signatureBuffer = signer.sign({ key: keyObject, dsaEncoding: 'der' });
    const signature = signatureBuffer.toString('hex');

    return {
      statusCode: 200,
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({ payload, signature }),
    };
  } catch (err) {
    return {
      statusCode: 500,
      body: `An error occurred: ${err.message}`,
    };
  }
};

File: netflify/functions/validate.py
import json
from datetime import datetime, timedelta
import os
from cryptography.hazmat.primitives import hashes, serialization
from cryptography.hazmat.primitives.asymmetric import ec
import dateutil.parser

def handler(event, context):
    try:
        # --- CONFIGURATION ---
        # The private key should be stored as an environment variable in Netlify
        private_key_pem = os.environ.get('VLTR_BETA_PRIVATE_KEY')
        if not private_key_pem:
            return {'statusCode': 500, 'body': 'Server configuration error: Private key not found.'}
            
        # The expiration date for the beta
        expiration_date = datetime(2026, 1, 1) # Year, Month, Day

        # --- TOKEN GENERATION ---
        private_key = serialization.load_pem_private_key(private_key_pem.encode(), password=None)
        
        # Create the data payload
        payload = {
            'issue_date': datetime.utcnow().isoformat(),
            'expiration_date': expiration_date.isoformat(),
            'issuer': 'VLTR Beta Program'
        }
        payload_bytes = json.dumps(payload, sort_keys=True).encode()

        # Sign the payload
        signature = private_key.sign(payload_bytes, ec.ECDSA(hashes.SHA256()))

        # Combine payload and signature into the final token
        token = {
            'payload': payload,
            'signature': signature.hex()
        }

        return {
            'statusCode': 200,
            'headers': {'Content-Type': 'application/json'},
            'body': json.dumps(token)
        }
    except Exception as e:
        return {
            'statusCode': 500,
            'body': f'An error occurred: {str(e)}'
        }